Assets/Scripts/Systems/Modules/ModulesInfo.cs

```diff
@@ -7,97 +7,169 @@
         }
 
-        public void RemoveModule(Module moduleToRemove)
+        public bool ModulePositionValid(Module module)
         {
-            foreach (Module module in Modules)
+            Coordinates rootPos = module.RootPosition;
+            foreach (Coordinates gridPos in module.Data.GridPositions)
             {
-                if (module == moduleToRemove)
+                if (rootPos.x + gridPos.x >= 0 && rootPos.x + gridPos.x < _columnLength &&
+                    rootPos.y + gridPos.y >= 0 && rootPos.y + gridPos.y < _rowHeight)
                 {
-                    int row = module.PivotPosition.y;
-                    int column = module.PivotPosition.x;
-                    
-                    foreach (Coordinates coords in module.Data.GridPositions)
+                    if (_grid[rootPos.y + gridPos.y, rootPos.x + gridPos.x] != null ||
+                        _slotGrid[rootPos.y + gridPos.y, rootPos.x + gridPos.x] == null ||
+                        (_slotGrid[rootPos.y + gridPos.y, rootPos.x + gridPos.x].ValidTypes & module.Data.Type) ==
+                        ModuleType.None)
                     {
-                        if (column + coords.x > 0 && column + coords.x < ColumnLength &&
-                            row + coords.y > 0 && row + coords.y < RowHeight)
-                        {
-                            Grid[row + coords.y, column + coords.x] = null;
-                        }
+                        return false;
                     }
                 }
             }
 
-            Modules.Remove(moduleToRemove);
+            return true;
         }
-        public string id => "module_grid";
+
         public object SaveState()
         {
             return new SaveData
             {
-                ModuleGrid = ModuleList.GridToLists(Grid),
-                Modules = Modules
+                GridHeight = _rowHeight,
+                GridWidth = _columnLength,
+                Modules = _modules,
+                ModuleSlots = _moduleSlots
             };
         }
 
         public void LoadState(JObject state)
         {
             var saveData = state.ToObject<SaveData>();
-            Grid = ModuleList.ListsToGrid(saveData.ModuleGrid);
-            Modules = saveData.Modules;
+            _rowHeight = saveData.GridHeight;
+            _columnLength = saveData.GridWidth;
+            _grid = new Module[_rowHeight, _columnLength];
+            _modules = saveData.Modules;
+            _moduleSlots = saveData.ModuleSlots;
+            
+            foreach (Module module in _modules)
+            {
+                AddModule(module);
+            }
+
+            foreach (ModuleSlot moduleSlot in _moduleSlots)
+            {
+                AddModuleSlot(moduleSlot);
+            }
         }
 
         [Serializable]
         private struct SaveData
         {
-            public List<ModuleList> ModuleGrid;
+            public int GridHeight;
+            public int GridWidth;
```

Comment:
Don't think this needs to be saved, since you get it from ship info in Awake (which will load the correct values from ship data when it is created)